explicit manual selection with `manually_selected_only`

LLFourn opened the pull request:
This makes it possible to choose a UTXO manually without having to choose them *all* manually. I introduced the `manually_selected_only` option to enforce that only manually selected UTXOs can be used.

I was implementing: https://github.com/bitcoindevkit/bdk/issues/114 and found that I it didn't make much sense without this feature. You want to be able to add a foreign utxo without turning on manual coin selection like `.add_utxo` does. I felt like it was too awkward to have `.add_utxo` turn on manual selection while my `.add_foreign_utxo` did not.
I think this feature makes sense in its own right as well.

Unfortunately, this test: https://github.com/bitcoindevkit/bdk/blob/master/src/wallet/mod.rs#L2101 fails and I'm not sure how to fix it or whether it should be fixed. I don't really understand what that test is testing. It asserts that it should fail with `InsufficientFunds` even though there are enough funds in the wallet to cover the fee bump. Help!

This is a draft until that is resolved and I've written tests/docs for the new feature. 

One thing I am slightly concerned about is that I have introduced the ability to redundantly do `add_utxo` and then do `send_all` which would have spend the utxo anyyway. Perhaps this should be a error/panic in tx_builder.

LLFourn committed:
[wallet] Replace ChangeSpendPolicy::filter_utxos with a predicate

To make composing it with other filtering conditions easier.

LLFourn committed:
[wallet] Make 'unspendable' into a HashSet

to avoid awkwardly later on.

afilini commented:
Yeah I think it makes sense as a feature to just be able to pick some utxos and then letting bdk do the coin selection to cover everything else.

I haven't read the code yet, I just wanted to explain what that test tries to do: basically the wallet receives two incoming transactions (one from `get_funded_wallet()` and then one more right afterwards). Then it creates a tx that manually picks only one of the two utxos and that is `send_all`, meaning it has no change. Maybe the confusion comes from the fact that `send_all` doesn't in fact "empty" the wallet, if there are some specific utxos selected. It just spends all the selected ones and sends everything (minus the fee) to the address specified, without creating the change. Then it tries to bump the fees of this transaction, to a very high fee rate that pretty much consumes the whole output and only leaves a dust amount in there. Since again there's only one utxo specified manually, `bump_fee()` doesn't try to add more inputs and only sticks to that one. This forces the wallet to fail, because at that high fee rate and being unable to add extra inputs it can't create an output large enough to be non-dust. `InsufficientFunds` is probably not a great error here, I think there was a TODO in the code to add an extra error like `OutputBelowDustLimit` for cases like this. The problem is that sometimes having an output below dust limit actually means that you don't have enough funds, it depends on whether you are allowing the wallet to use all the utxos or not: when you specify manually the utxos it means that those utxos cannot cover the outputs/fee rate that you requested, not that you don't have enough funds in the whole wallet.

RCasatta commented on src/wallet/mod.rs:
```diff
@@ -972,38 +957,53 @@ where
+        manually_selected: &[OutPoint],
+        send_all: bool,
+        manual_only: bool,
+    ) -> Result<(Vec<(UTXO, usize)>, Vec<(UTXO, usize)>), Error> {
```
Did you consider introducing an enum like 
```
enum UseUTXO {
  May(UTXO, usize),
  Must(UTXO, usize),
}
```
so that the Result would be just `Vec<UseUTXO>` ?

LLFourn replied:
`coin_select` actually uses the two distinct vectors. It sorts the `may_use` in order of size before selecting them. `must_use` are not sorted. So putting them altogether in the same vector would complicate that a bit so I think it's best to leave it as it is now at least for this PR.

LLFourn commented:
@afilini Thanks. I think I get you. The source of my confusion seems to be the different semantics of `send_all` in `create_tx` vs `bump_fee`. In `create_tx` it not only means send all selected inputs to an output, it means select all your inputs and send them all there!

I think this can be made more consistent by simply having `drain_wallet` as a separate option and making `send_all` just mean send everything that is selected to this output. Of course it would then be an error to use `send_all` without turning on either `manually_selected_only` or `drain_wallet` in `create_tx`.

LLFourn committed:
[wallet] Stop implicitly enforcing manaul selection by .add_utxo

This makes it possible to choose a UTXO manually without having to
choose them *all* manually. I introduced the `manually_selected_only`
option to enforce that only manually selected utxos can be used.

To stop the cli semantics changing I made the `utxos` keep the old
behaviour by calling `manually_selected_only`.

LLFourn commented on src/wallet/mod.rs:
```diff
@@ -595,30 +592,29 @@ where
+        let mut must_use_utxos =
+            rbf::filter_available(self.database.borrow().deref(), must_use_utxos.into_iter())?;
+        let may_use_utxos =
+            rbf::filter_available(self.database.borrow().deref(), may_use_utxos.into_iter())?;
```
💡  This bit can maybe be done inside `get_must_may_use_utxos` once I am certain I understand its purpose.

afilini replied:
Yeah, I think it should be possible to move it in there and only do one filtering before the utxos are split between "may" and "must" use

afilini replied:
Ah no, wait a second: you should only do the filtering on `bump_fee`, not `create_tx`. So maybe moving it in there is not a good idea, because then you need an extra parameter to tell whether you are creating or bumping a tx.

LLFourn reviewed (commented):
This is ready for review now :).

Each commit can be reviewed individually -- I had to clean up a couple of things to get it right.

🚨  NOTE: this change is quite spicy since `.add_utxo` and then `send_all` now **sends all your funds** since `.add_utxo` no longer implies manual coin selection (i.e.  `add_utxo`  then `send_all` is redundant). I think this change is acceptable because the fact that `send_all` would send all your funds was not documented before this PR except on the cli as far as I can tell. I've preserved the api of `send_all` on the cli.

I feel like the right solution (for another PR) is to split `send_all` into two options:

- `set_single_recipient(script_pubkey)` : this recipient will get all selected funds and no change will be added.
- `drain_wallet`: modulo other filters like coin selection and unspendable this will select all the inputs owned by the wallet for spending.

`drain_wallet` might be useful in its own right so you can make wallet that spends all its funds in every transaction but puts the excess into its change each time (reduces UTXOs)!

afilini commented:
> @afilini Thanks. I think I get you. The source of my confusion seems to be the different semantics of `send_all` in `create_tx` vs `bump_fee`. In `create_tx` it not only means send all selected inputs to an output, it means select all your inputs and send them all there!

Yes, `send_all` on `bump_fee` is currently only used to basically signal that you can update the only output that the tx has, because otherwise if the scriptPubKey is not owned by the wallet it would leave that output untouched, add more inputs and add a change. The idea was that if you specify `send_all` when you create the tx, you should also do it when you bump the fees, otherwise you might get unexpected results.

> :rotating_light: NOTE: this change is quite spicy since .add_utxo and then send_all now sends all your funds since .add_utxo no longer implies manual coin selection (i.e. add_utxo then send_all is redundant). I think this change is acceptable because the fact that send_all would send all your funds was not documented before this PR except on the cli as far as I can tell. I've preserved the api of send_all on the cli.

That's not really a problem IMHO, as you said those were pretty much undocumented "quirks" of the code, and we are not in a stable "phase" yet, so I don't think people are expecting stability in that sense. Maybe we should start writing a changelog for all those updates though, that would help people experimenting with it keep track of what we change.

> I feel like the right solution (for another PR) is to split send_all into two options:
>  * set_single_recipient(script_pubkey) : this recipient will get all selected funds and no change will be added.
>  * drain_wallet: modulo other filters like coin selection and unspendable this will select all the inputs owned by the wallet for spending.

Agreed, I like this idea. Not sure if it would be better to call it `single_recipient` or `no_change`, meaning that I don't know if it's better to name the feature with "what it does" (setting a single recipient) vs "how it does it" (doesn't add a change).. but overall I like the idea of splitting up `send_all` into two distinct things.

I was also thinking about adding a generic to `TxBuilder` using `PhantomData` that we could use to differentiate the builders that go to `create_tx` and the builders that go to `bump_fee`. I'm pretty sure there are some options that are currently ignored by `bump_fee`, it would be nice if the compiler also prevented you from setting them when you are making a builder for a fee bump.

afilini commented:
The code looks good to me, the only suggestion I would make is that you could do the same refactoring you did on `ChangeSpendPolicy::filter_utxos` to `rbf::filter_available`, but that's not a requiremente and we can also do it in a separate PR.

Other than that, you should rebase on `master` and then we are ready to merge this :)

LLFourn committed:
Incorporate RBF rules into utxo selection function

LLFourn commented:
Done. I deleted `rbf.rs` since it didn't seem to be doing much and moved the logic into may_must_utxos function.

I made [this issue](https://github.com/bitcoindevkit/bdk/issues/144.) because I only apply this filter to `may_use` utxos. I guess it should be an error to manually add an unconfirmed output to a fee_bump.

Another related issue is that by default I think that you should only spend confirmed coins? This doesn't seem to be the case at the moment. There's an argument that coin selection should handle this but I think that an option in txbuilder to only use outputs that have at least X confirmations with X being 1 by default would make sense.

afilini commented:
> I made this issue because I only apply this filter to may_use utxos. I guess it should be an error to manually add an unconfirmed output to a fee_bump.

Is there any particular reason why you didn't also apply the filter to `must_use` utxos? I guess if the user picks an unconfirmed utxo we could either return a `UnknownUTXO` or maybe add a specific error for this case where the utxo actually exist, but it's not confirmed yet. That would also be useful for when we add the "ConfirmationSpendPolicy" (see below).

> Another related issue is that by default I think that you should only spend confirmed coins? This doesn't seem to be the case at the moment. There's an argument that coin selection should handle this but I think that an option in txbuilder to only use outputs that have at least X confirmations with X being 1 by default would make sense.

That's true, I wanted to add a parameter to `TxBuilder` similar to the change policy but for confirmations, where you can choose between "only confirmed", "allow unconfirmed change", etc. I got stuck with that because I also wanted to add options specifically for incoming rbf transactions, like "spend unconfirmed only if it's not rbf". The problem is that finding out whether a tx is replaceable is very tricky, because it could also not explicitly signal replaceability but inherit it from previous transaction.

So yeah, we should definitely have that, we just have to figure out a nice way of detecting rbf transactions, or maybe just decide to ignore that part for now and only look at whether a tx is confirmed or not.

LLFourn commented:
> Is there any particular reason why you didn't also apply the filter to must_use utxos?

Because they're `must_use`! I must use them otherwise my variable name is confusing :P. I can re-apply the filter to `must_use` as well if that's preferable? Instead I thought it might be ok to leave the footgun there and make an issue to fix it and test it with it's own specific error.

afilini commented:
Yeah, now that I think about it there are a few inconsistencies with how "conflicts" are resolved: for instance, if you mark something as "unspendable", but then add it as a "must spend", it will be spend. Also if you specify a "ChangeSpendPolicy" but then add a utxo that doesn't satisfy that policy, it will still spend it. So returning an error here would probably introduce an inconsistency with those other stuff.

What you've done is probably the right thing here, keep the current behavior (even if it's not the best) but open an issue to keep that in mind and potentially change it in the future.
